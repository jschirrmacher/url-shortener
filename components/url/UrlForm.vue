<script setup lang="ts">
interface Props {
  shortUrl?: string
  shortCode?: string
  originalUrl?: string
  title?: string
  owner?: string
}

interface Emits {
  (e: "changed", data: { shortCode: string; originalUrl: string; title: string; owner?: string }): void
  (e: "cancel"): void
}

const props = withDefaults(defineProps<Props>(), {
  title: '',
  shortCode: '',
  shortUrl: '',
  originalUrl: '',
  owner: ''
})

const emit = defineEmits<Emits>()

const { copyToClipboard } = useClipboard()
const { user: currentUser } = useAuth()
const usersStore = useUsersStore()

const isEditMode = computed(() => !!props.shortUrl)
const isAdmin = computed(() => currentUser.value?.role === 'admin')

const formData = reactive({
  originalUrl: "",
  shortCode: "",
  title: "",
  owner: ""
})

// Initialize form data when component mounts or props change
watchEffect(() => {
  formData.originalUrl = props.originalUrl || ""
  formData.shortCode = props.shortCode || ""
  formData.title = props.title || ""
  formData.owner = props.owner || currentUser.value?.username || ""
})

// Load users for admin
onMounted(() => {
  if (isAdmin.value) {
    usersStore.loadUsers()
  }
})

// Validation state
const validationErrors = reactive({
  originalUrl: "",
  shortCode: ""
})

// Validation
function isValidUrl(url: string): boolean {
  try {
    new URL(url)
    return true
  } catch {
    return false
  }
}

function isValidShortCode(code: string): boolean {
  if (!code) return true
  if (code.length < 6) return false
  return /^[a-zA-Z0-9_\-]+$/v.test(code)
}

function validateForm(): boolean {
  // Trigger validation for empty fields
  if (!formData.originalUrl.trim()) {
    validationErrors.originalUrl = "URL ist erforderlich"
  }

  // Return true if no validation errors exist
  return !validationErrors.originalUrl && !validationErrors.shortCode
}

// Real-time validation
watch(() => formData.originalUrl, (value) => {
  if (value.trim() && !isValidUrl(value)) {
    validationErrors.originalUrl = "Ungültige URL"
  } else {
    validationErrors.originalUrl = ""
  }
})

watch(() => formData.shortCode, (value) => {
  if (value) {
    if (value.length < 6) {
      validationErrors.shortCode = "Mindestens 6 Zeichen erforderlich"
    } else if (!/^[a-zA-Z0-9_\-]+$/v.test(value)) {
      validationErrors.shortCode = "Nur Buchstaben, Zahlen, Bindestriche und Unterstriche erlaubt"
    } else {
      validationErrors.shortCode = ""
    }
  } else {
    validationErrors.shortCode = ""
  }
})

const isFormValid = computed(() => {
  const hasUrl = formData.originalUrl.trim() !== ''
  const validUrl = hasUrl && isValidUrl(formData.originalUrl)
  const validShortCode = isValidShortCode(formData.shortCode)
  return hasUrl && validUrl && validShortCode
})

// Handle form submission
function handleSubmit() {
  if (!validateForm()) return

  emit('changed', {
    shortCode: formData.shortCode,
    originalUrl: formData.originalUrl,
    title: formData.title,
    owner: formData.owner
  })
}

function handleCancel() {
  emit('cancel')
}
</script>

<template>
  <div class="url-form">
    <!-- Edit mode: Show current short URL -->
    <div v-if="isEditMode && shortUrl" class="readonly-field">
      <label>Kurz-URL</label>
      <div class="input-with-button">
        <input :value="shortUrl" readonly class="readonly-input">
        <BaseButton variant="secondary" size="sm" @click="copyToClipboard(shortUrl)">
          Kopieren
        </BaseButton>
      </div>
    </div>

    <form @submit.prevent="handleSubmit">
      <!-- URL field (first and most important) -->
      <div class="form-field">
        <label for="originalUrl">{{ isEditMode ? 'Ziel-URL' : 'URL zum Verkürzen' }}</label>
        <input
          id="originalUrl"
          v-model="formData.originalUrl"
          type="url"
          required
          placeholder="https://example.com"
          :class="{ 'error': validationErrors.originalUrl }"
        >
        <p v-if="validationErrors.originalUrl" class="error-message">
          {{ validationErrors.originalUrl }}
        </p>
      </div>

      <!-- Optional fields in responsive layout -->
      <div class="optional-fields">
        <!-- Short code field (smaller) -->
        <div class="form-field shortcode-field">
          <label for="shortCode">{{ isEditMode ? 'Kurz-Code' : 'Kurz-Code (optional)' }}</label>
          <input
            id="shortCode"
            v-model="formData.shortCode"
            type="text"
            :placeholder="isEditMode ? 'z.B. mein-link' : 'Auto-generiert'"
            :required="isEditMode"
            pattern="[a-zA-Z0-9_\-]{6,}"
            title="Mindestens 6 Zeichen: Buchstaben, Zahlen, Bindestriche und Unterstriche erlaubt"
            :class="{ 'error': validationErrors.shortCode }"
          >
          <p v-if="validationErrors.shortCode" class="error-message">
            {{ validationErrors.shortCode }}
          </p>
        </div>

        <div v-if="isAdmin && usersStore.users.length > 0" class="form-field owner-field" data-testid="owner-field">
          <label for="owner">Eigentümer</label>
          <select id="owner" v-model="formData.owner" class="owner-select">
            <option 
              v-for="user in usersStore.users" 
              :key="user.username"
              :value="user.username"
            >
              {{ user.username }}
            </option>
          </select>
        </div>
      </div>

      <div class="form-field title-field">
        <label for="title">Titel (optional)</label>
        <input id="title" v-model="formData.title" type="text" placeholder="Beschreibender Titel für den Link">
      </div>

      <!-- Buttons -->
      <div class="form-actions">
        <BaseButton variant="primary" type="submit" :disabled="!isFormValid">
          {{ isEditMode ? 'Speichern' : 'URL verkürzen' }}
        </BaseButton>

        <BaseButton v-if="isEditMode" variant="secondary" type="button" @click="handleCancel">
          Abbrechen
        </BaseButton>
      </div>
    </form>
  </div>
</template>

<style scoped>
.url-form {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.readonly-field {
  margin-bottom: 1rem;
}

.readonly-field label {
  display: block;
  font-size: 0.875rem;
  font-weight: 500;
  color: #374151;
  margin-bottom: 0.5rem;
}

.input-with-button {
  display: flex;
  gap: 0.5rem;
}

.readonly-input {
  flex: 1;
  padding: 0.5rem;
  border: 1px solid #d1d5db;
  border-radius: 0.375rem;
  background-color: #f9fafb;
  color: #6b7280;
  font-size: 0.875rem;
}

.form-field {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  margin-top: 1rem;
}

.form-field label {
  font-size: 0.875rem;
  font-weight: 500;
  color: #374151;
  padding-left: 0.25rem;
}

.form-field input {
  padding: 0.75rem;
  border: 1px solid #d1d5db;
  border-radius: 0.375rem;
  font-size: 1rem;
}

.form-field input:focus {
  outline: none;
  border-color: #3b82f6;
  box-shadow: 0 0 0 3px rgba(59, 130, 246, 0.1);
}

.form-field input.error {
  border-color: #ef4444;
}

.form-field input.error:focus {
  border-color: #ef4444;
  box-shadow: 0 0 0 3px rgba(239, 68, 68, 0.1);
}

.error-message {
  font-size: 0.75rem;
  color: #ef4444;
  margin: 0;
}

@media (prefers-color-scheme: dark) {
  .error-message {
    color: #f87171;
  }
}

.field-hint {
  font-size: 0.75rem;
  color: #6b7280;
  margin: 0;
  padding-left: 0.25rem;
}

.field-hint.error-hint {
  color: #dc2626;
}

.optional-fields {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

@media (min-width: 768px) {
  .optional-fields {
    flex-direction: row;
    gap: 1rem;
  }

  .shortcode-field {
    flex: 0 0 200px;
    /* Fixed width for short code */
  }

  .owner-field {
    flex: 0 0 200px;
    /* Fixed width for owner selection */
  }
}

.form-actions {
  display: flex;
  gap: 0.75rem;
  margin-top: 1rem;
}

.owner-select {
  width: 100%;
  padding: 0.75rem;
  border: 1px solid var(--border-primary);
  border-radius: 0.375rem;
  font-size: 0.875rem;
  background-color: var(--bg-primary);
  color: var(--text-primary);
}

.owner-select:focus {
  outline: none;
  border-color: #2563eb;
  box-shadow: 0 0 0 2px rgba(37, 99, 235, 0.2);
}
</style>
