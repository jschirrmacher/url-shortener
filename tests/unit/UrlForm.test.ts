/**
 * @vitest-environment happy-dom
 */
import { describe, it, expect, vi, beforeEach } from "vitest"
import { mount } from "@vue/test-utils"
import { ref, reactive, computed, watch, watchEffect } from "vue"
import UrlForm from "~/components/url/UrlForm.vue"

// Mock useClipboard
vi.stubGlobal('useClipboard', () => ({
  copyToClipboard: vi.fn()
}))

// Default mocks
const mockUser = ref({ role: 'user', username: 'testuser' })
const mockUsers = ref([])

vi.stubGlobal('useAuth', () => ({
  user: mockUser
}))

vi.stubGlobal('useUsersStore', () => ({
  users: mockUsers,
  loadUsers: vi.fn()
}))

// Mock Vue composables
vi.stubGlobal('ref', ref)
vi.stubGlobal('reactive', reactive)
vi.stubGlobal('computed', computed)
vi.stubGlobal('watch', watch)
vi.stubGlobal('watchEffect', watchEffect)
vi.stubGlobal('onMounted', vi.fn())

describe("UrlForm Component", () => {
  beforeEach(() => {
    // Reset to default user
    mockUser.value = { role: 'user', username: 'testuser' }
    mockUsers.value = []
  })
  it("should match snapshot in create mode", () => {
    const wrapper = mount(UrlForm, {
      global: {
        stubs: {
          BaseButton: true,
          AlertMessage: true
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should match snapshot in edit mode", () => {
    const wrapper = mount(UrlForm, {
      props: {
        shortUrl: "http://localhost:3000/abc123",
        shortCode: "abc123",
        originalUrl: "https://example.com",
        title: "Example Website"
      },
      global: {
        stubs: {
          BaseButton: true,
          AlertMessage: true
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should emit changed event with form data", async () => {
    const wrapper = mount(UrlForm, {
      global: {
        stubs: {
          BaseButton: true,
          AlertMessage: true
        }
      }
    })

    // Fill form
    await wrapper.find('#originalUrl').setValue('https://test.com')
    await wrapper.find('#shortCode').setValue('test123')
    await wrapper.find('#title').setValue('Test Title')

    // Submit form
    await wrapper.find('form').trigger('submit.prevent')

    // Check emit
    const emitted = wrapper.emitted('changed')
    expect(emitted).toBeTruthy()
    expect(emitted![0]).toEqual([{
      originalUrl: 'https://test.com',
      shortCode: 'test123',
      title: 'Test Title',
      owner: 'testuser'
    }])
  })

  it("should show cancel button in edit mode", () => {
    const wrapper = mount(UrlForm, {
      props: {
        shortUrl: "http://localhost:3000/abc123"
      },
      global: {
        stubs: {
          BaseButton: true,
          AlertMessage: true
        }
      }
    })

    // Should have more buttons in edit mode (including copy button)
    const buttons = wrapper.findAllComponents({ name: 'BaseButton' })
    expect(buttons.length).toBeGreaterThan(1)
  })

  it("should not show owner selection for regular users", () => {
    const wrapper = mount(UrlForm, {
      global: {
        stubs: {
          BaseButton: true,
          AlertMessage: true
        }
      }
    })

    // Owner selection should not be visible (regular user)
    const ownerField = wrapper.find('[data-testid="owner-field"]')
    expect(ownerField.exists()).toBe(false)
  })

  it("should include owner field in emitted data when user is admin", async () => {
    // This test verifies that admin users can set owner field
    const wrapper = mount(UrlForm, {
      global: {
        stubs: {
          BaseButton: true,
          AlertMessage: true
        }
      }
    })

    // Fill form data
    await wrapper.find('#originalUrl').setValue('https://test.com')
    await wrapper.find('#shortCode').setValue('test123')
    await wrapper.find('#title').setValue('Test Title')

    // Submit form
    await wrapper.find('form').trigger('submit.prevent')

    // Check emitted data includes owner field (from mock user)
    const emitted = wrapper.emitted('changed')
    expect(emitted).toBeTruthy()
    const eventData = emitted![0]![0] as { owner: string }
    expect(eventData).toHaveProperty('owner')
    expect(eventData.owner).toBe('testuser')
  })
})
