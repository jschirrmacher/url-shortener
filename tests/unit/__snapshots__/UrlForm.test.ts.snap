// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UrlForm Component > should match snapshot in create mode 1`] = `
"<div data-v-6bb9e226="" class="url-form">
  <!-- Edit mode: Show current short URL -->
  <!--v-if-->
  <form data-v-6bb9e226="">
    <!-- URL field (first and most important) -->
    <div data-v-6bb9e226="" class="form-field"><label data-v-6bb9e226="" for="originalUrl">URL zum Verkürzen</label><input data-v-6bb9e226="" id="originalUrl" type="url" required="" placeholder="https://example.com" class="">
      <!--v-if-->
    </div><!-- Optional fields in responsive layout -->
    <div data-v-6bb9e226="" class="optional-fields">
      <!-- Short code field (smaller) -->
      <div data-v-6bb9e226="" class="form-field shortcode-field"><label data-v-6bb9e226="" for="shortCode">Kurz-Code (optional)</label><input data-v-6bb9e226="" id="shortCode" type="text" placeholder="Auto-generiert" pattern="[a-zA-Z0-9_\\-]{6,}" title="Mindestens 6 Zeichen: Buchstaben, Zahlen, Bindestriche und Unterstriche erlaubt" class="">
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
    <div data-v-6bb9e226="" class="form-field title-field"><label data-v-6bb9e226="" for="title">Titel (optional)</label><input data-v-6bb9e226="" id="title" type="text" placeholder="Beschreibender Titel für den Link"></div><!-- Buttons -->
    <div data-v-6bb9e226="" class="form-actions">
      <base-button-stub data-v-6bb9e226="" variant="primary" type="submit" disabled="true"></base-button-stub>
      <!--v-if-->
    </div>
  </form>
</div>"
`;

exports[`UrlForm Component > should match snapshot in edit mode 1`] = `
"<div data-v-6bb9e226="" class="url-form">
  <!-- Edit mode: Show current short URL -->
  <div data-v-6bb9e226="" class="readonly-field"><label data-v-6bb9e226="">Kurz-URL</label>
    <div data-v-6bb9e226="" class="input-with-button"><input data-v-6bb9e226="" readonly="" class="readonly-input" value="http://localhost:3000/abc123">
      <base-button-stub data-v-6bb9e226="" variant="secondary" size="sm"></base-button-stub>
    </div>
  </div>
  <form data-v-6bb9e226="">
    <!-- URL field (first and most important) -->
    <div data-v-6bb9e226="" class="form-field"><label data-v-6bb9e226="" for="originalUrl">Ziel-URL</label><input data-v-6bb9e226="" id="originalUrl" type="url" required="" placeholder="https://example.com" class="">
      <!--v-if-->
    </div><!-- Optional fields in responsive layout -->
    <div data-v-6bb9e226="" class="optional-fields">
      <!-- Short code field (smaller) -->
      <div data-v-6bb9e226="" class="form-field shortcode-field"><label data-v-6bb9e226="" for="shortCode">Kurz-Code</label><input data-v-6bb9e226="" id="shortCode" type="text" placeholder="z.B. mein-link" required="" pattern="[a-zA-Z0-9_\\-]{6,}" title="Mindestens 6 Zeichen: Buchstaben, Zahlen, Bindestriche und Unterstriche erlaubt" class="">
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
    <div data-v-6bb9e226="" class="form-field title-field"><label data-v-6bb9e226="" for="title">Titel (optional)</label><input data-v-6bb9e226="" id="title" type="text" placeholder="Beschreibender Titel für den Link"></div><!-- Buttons -->
    <div data-v-6bb9e226="" class="form-actions">
      <base-button-stub data-v-6bb9e226="" variant="primary" type="submit" disabled="false"></base-button-stub>
      <base-button-stub data-v-6bb9e226="" variant="secondary" type="button"></base-button-stub>
    </div>
  </form>
</div>"
`;
